packages/shared/ReactTypes.js

```diff
@@ -174,3 +174,21 @@ export type ReactCustomFormAction = {
   target?: string,
   data?: null | FormData,
 };
+
+// Linked list representing the identity of a component given the component/tag
+// name and key. The name might be minified but we assume that it's going to be
+// the same generated name. Typically because it's just the same compiled output
+// in practice.
+export type KeyNode = [
+  null | KeyNode /* parent */,
+  string | null /* name */,
+  string | number /* key */,
+];
+
+// This is an opaque type returned by decodeFormState on the server, but it's
+// defined in this shared file because the same type is used by React on
+// the client.
```

Comment:
Seems like we should just pass the string value through. This shouldn't really be shared with the client since it doesn't know the format and the format could change to just a string for example. It might not even be JSON. I think of the key as just a string similar to a key on a React Element.

Reply:
It also indicates that there's not parity since Flight does JSON.parse and Fizz does JSON.stringify. So there's not one owner. Seems like Fizz should be the sole owner and do the JSON.parse part and while passing around it's just a string. This means that in the RSC protocol it'll be encoded twice as a JSON string inside of JSON but really it's just an id. Could be a hash too.

Reply:
I guess ReactFormState<S> would still be an opaque type here but KeyNode can go away and live in Fizz.